### ISSUE TITLE
Default scraping of zh.wiki content should be mainland traditional

### ISSUE THREAD

**Issue by Popolechien** [openzim/mwoffliner#840]
There are several versions of Chinese cohabiting on zh.wikipedia (mainland simplified, HongKong traditional, Macao traditional, Malaysian simplified, Singapore simplified, Taiwan traditional). The wiki switches easily from on to the other as only characters are impacted (not grammar).

I'm not entirely sure how this works on our end but from many reports I get we're not serving people with mainland traditional (大陆简体) content, which is both easier to use and has a lot more readers. It would be nice, therefore, to scrape it correctly. I also remember that this very issue was raised a while back, but I could not find it anywhere.

**Comment by kelson42** [comment]
@Popolechien In which flavour is written the wiki code on zh.wikipedia?

**Comment by Popolechien** [comment]
If it's anything like German, in the language used by the initial editor. But as I said, you can change settings so as to make it default to read in one script or the other and never know who wrote what in which language. With our latin alphabet the closest comparison I can think of would be to change fonts or replace automatically the German "ß" with the Swiss "ss".

**Comment by Popolechien** [comment]
@fantasticfears might be able to help.

**Comment by kelson42** [comment]
@Popolechien OK, so we have different flavours in the wiki code, its autodetected and then displayed like the user wants?

**Comment by Popolechien** [comment]
yup.

**Comment by kelson42** [comment]
@Popolechien So what is currently the flavour of WPZH ZIM file?

**Comment by Popolechien** [comment]
Just checked and the last ZIM we (zh_all_2019_06) have seems to be mainland simplified - at this stage we should simply wait for a native user/speaker to clarify things.

**Comment by fantasticfears** [comment]
This is called language variants. The conversion process basically includes script conversion and word conversion. I tried `wikipedia_zh_basketball_nopic_2019-06.zim` and believe it's rendered from the source script so it's not what people see online. Does kiwix grab the source script and render it locally?

**Comment by fantasticfears** [comment]
@ISNIT0 Maybe Joe has some recommendations?

**Comment by kelson42** [comment]
@fantasticfears Like written a long time ago in the other ticket, language variants are not supported by Parsoid. So for now, there is nothing we can do.

**Comment by fantasticfears** [comment]
That's sad but the rendering engine from WMF is switching to PHP. How much work we have to use PHP parser?

**Comment by kelson42** [comment]
@fantasticfears If Parsoide/MCS moves to PHP, that will probably need no or only a few work on our side as the API is table. But all of this is in the future, and we have no control on this, see upstream ticket in earlier comment.

**Comment by fantasticfears** [comment]
Original ticket for language variants is here: https://phabricator.wikimedia.org/T43716.
The PHP parsoid is ongoing but the language variants will be the other work but we can check that out later for less work.

**Comment by fantasticfears** [comment]
Happy new year. It's time to restart the topic. Parsoid-PHP is deployed. But that doesn't move things for us entirely. Now I examined the downloader and I think visualeditor won't expose something we need but mobile-section does the trick. `Accept-Language` needs to be set to `zh-cn` to get our dump.

Now the returned value is a scramble.

```
<p data-mw-variant-lang=\"zh-cn\">可<span data-mw-variant-lang=\"zh-hans\" data-mw-variant=\"{&quot;twoway&quot;:[{&quot;l&quot;:&quot;zh-hans&quot;,&quot;t&quot;:&quot;說&quot;},{&quot;l&quot;:&quot;zh-cn&quot;,&quot;t&quot;:&quot;说&quot;}],&quot;rt&quot;:true}\">说</span>是一<span data-mw-variant-lang=\"zh-hans\" data-mw-variant=\"{&quot;twoway&quot;:[{&quot;l&quot;:&quot;zh-hans&quot;,&quot;t&quot;:&quot;種&quot;},{&quot;l&quot;:&quot;zh-cn&quot;,&quot;t&quot;:&quot;种&quot;}],&quot;rt&quot;:true}\">种</span>衍生自<span data-mw-variant-lang=\"zh-hans\" data-mw-variant=\"{&quot;twoway&quot;:[{&quot;l&quot;:&quot;zh-hans&quot;,&quot;t&quot;:&quot;親&quot;},{&quot;l&quot;:&quot;zh-cn&quot;,&quot;t&quot;:&quot;亲&quot;}],&quot;rt&quot;:true}\">亲</span>人之<span data-mw-variant-lang=\"zh-hans\" data-mw-variant=\"{&quot;twoway&quot;:[{&quot;l&quot;:&quot;zh-hans&quot;,&quot;t&quot;:&quot;間&quot;},{&quot;l&quot;:&quot;zh-cn&quot;,&quot;t&quot;:&quot;间&quot;}],&quot;rt&quot;:true}\">间</span>的<span data-mw-variant-lang=\"zh-hans\" data-mw-variant=\"{&quot;twoway&quot;:[{&quot;l&quot;:&quot;zh-hans&quot;,&quot;t&quot;:&quot;強&quot;},{&quot;l&quot;:&quot;zh-cn&quot;,&quot;t&quot;:&quot;强&quot;}],&quot;rt&quot;:true}\">强</span>烈<span data-mw-variant-lang=\"zh-hans\" data-mw-variant=\"{&quot;twoway&quot;:[{&quot;l&quot;:&quot;zh-hans&quot;,&quot;t&quot;:&quot;關愛&quot;},{&quot;l&quot;:&quot;zh-cn&quot;,&quot;t&quot;:&quot;关爱&quot;}],&quot;rt&quot;:true}\">关爱</span>、<a href=\"/wiki/忠诚\" title=\"忠诚\" class=\"mw-redirect\" data-mw-variant-orig=\"忠誠\">忠<span data-mw-variant-lang=\"zh-hans\" data-mw-variant=\"{&quot;twoway&quot;:[{&quot;l&quot;:&quot;zh-hans&quot;,&quot;t&quot;:&quot;誠&quot;},{&quot;l&quot;:&quot;zh-cn&quot;,&quot;t&quot;:&quot;诚&quot;}],&quot;rt&quot;:true}\">诚</span></a>及善意的情感
```

Mind a PR and some cleansing code at downloader?

**Comment by danielzgtg** [comment]
It is not a good idea to just convert everything to Simplified Chinese.

First of all, such conversion is lossy and irreversible. For example, both 發 (fa1, to send) and 髮 (fa3, hair) simplify to 发. It would be hard to recover this information if we do this.

Simplification is also a very sensitive political topic, and it would be better to not risk offending or excluding anybody. By leaving the characters as-is, we can tell anyone who complains to look at the source article, no matter which side they support.

I recommend sticking to the original suggestion in the SourceForge ticket. That would be to not do any conversion in mwoffliner but to add a converter to each client. This would be consistent to what the online zh.wikipedia.org is doing. It would give users the freedom to choose and not impose a decision on them.

**Comment by fantasticfears** [comment]
> I recommend sticking to the original suggestion in the SourceForge ticket. That would be to not do any conversion in mwoffliner but to add a converter to each client. This would be consistent to what the online zh.wikipedia.org is doing. It would give users the freedom to choose and not impose a decision on them.

No, it's not how zh.wikipedia.org works. zhwiki_p works with huge sets of manually defined rules with prefix matching algorithm to convert original articles into the expected language variants. This is not perfect and mistakes happen sometimes. Nevertheless, it works for more than a decade now.

> First of all, such conversion is lossy and irreversible. For example, both 發 (fa1, to send) and 髮 (fa3, hair) simplify to 发. It would be hard to recover this information if we do this.

As you can see from the snippet, the conversion rule is defined in `data-mw-variant`. There are rules to dealing with this type of errors. It's however manually controlled on zhwiki_p.

**Comment by danielzgtg** [comment]
> Nevertheless, it works for more than a decade now.

Ok, if it works. I'm fine with this if I can still read Kiwix in Traditional Chinese (臺灣正體).

**Comment by fantasticfears** [comment]
What you are reading now from Kiwix is perhaps the source. By implementing the feature I propose, you should be able to read the correct variant as you intend, for example, Traditional Chinese. But I notice that the issue's title is different than what I propose. What I want to achieve is to have at least two language variants dump for zhwiki.

**Comment by Popolechien** [comment]
@fantasticfears Feel free to rename the issue if you feel that your solution solves it AND doesn't leave Traditional Chinese out the window (which would definitely be best). Generally speaking, it would probably be better to simply rename it to "zh.wikipedia should also be accessible in simplified chinese" or something similar.

**Comment by danielzgtg** [comment]
@Popolechien Is there a typo in the issue body?

Title is like this:

> Default scraping of zh.wiki content should be mainland simplified characters

Do you mean simplified instead of traditional here (traditional is discouraged in the mainland)?:

> […] not serving people with mainland traditional […]

The Hanzi say "mainland simplified":

> 大陸简体

You are saying simplified in ⅔ of the places, but traditional in ⅓ of the places.

**Comment by Popolechien** [comment]
@danielzgtg To clarify: I opened the ticket with the intent of having articles available in 简体字 (simplified) instead of 繁体字 (traditional) everywhere, which seems to be the current default setting.

**Comment by fantasticfears** [comment]
@danielzgtg I can put some context as well. People in Mainland read Simplified Chinese with a variant setting (mainland simplified as you said). zhwiki works by providing 6 more localized variants based on regions. The source code can be either simplified and traditional. Or even a combination. These 6 variants use different localized words so they are manually mapped by a set of rules.
The pipeline roughly works like 1) convert to simplified/traditional based on variant setting 2) map words based on the dictionary. 

My intent would be to provide simplified and traditional (based on Taiwan variant) feature. If additional variants are asked by readers, it will be as easy as adding config. So this differs from what @Popolechien intends but my implementation should cover what you want. I don't want to make simplified version as default but to provide an option for people reads the simplified. IMO, there isn't a need to **make anything default**. It's not so hard for readers to choose what they want.

**Comment by dsd** [comment]
Just curious what is the current state of the zim file. From reading the above, I expected it to be in Traditional Chinese for Taiwan (which is incidentally what I'm looking for right now), with future plans to make the default be mainland/Simplified plus different regional variants available.

However I downloaded wikipedia_zh_top_mini_2020-05.zim and opened it in WebArchives. Looks like it is mainland/simplified. e.g. the page on England says
> 英国
> 大不列颠及北爱尔兰联合王国...

So that's using simplified chars e.g. 国 and 颠 (unlike the [zh-tw live version](https://zh.wikipedia.org/zh-tw/%E8%8B%B1%E5%9B%BD))

**Comment by Popolechien** [comment]
I'm looking at https://library.kiwix.org/wikipedia_zh_top_mini_2020-05 and looking for Germany, the US and other countries I see the title being simplified (德国,美国, etc.) and then the text in traditional (德意志聯邦共和**國**, 美利堅合眾**國**). Also, all articles on vehicles use 車 (trad.), not 车 (simpl.)

Seems like there's a lot of mish-mash of both scripts.

**Comment by dsd** [comment]
You're right. Thanks for clarifying. I didn't spot the mishmash earlier due to my low familiarity with simplified characters.

**Comment by dsd** [comment]
Trying to iterate on @erickguan's suggestion, I see that Parsoid is working well:

```
wget --header="Accept-Language: zh-cn" "https://zh.wikipedia.org/api/rest_v1/page/html/%E8%8B%B1%E5%9B%BD"
```
This gets a mainland Chinese version (simplified), and if I change to `Accept-Language: zh-tw` I get a Taiwan version (traditional), I checked by opening the downloaded HTML in a browser. (The <title> is using simplified in both cases but the content is good)

So is the next step here just as simple as adding a --accept-language parameter to mwoffliner, then insert the requested language into the HTTP headers in Downloader.ts `getRequestOptionsFromUrl`?

**Comment by Popolechien** [comment]
@kelson42 @bakshiutkarsha this is way out of my league, but would the above fix make sense to you?

**Comment by stevenling** [comment]
I think can  support Simplified Chinese and Traditional Chinese.

**Comment by Markus-Rost** [comment]
With the ActionParse render, we should be getting the default variant correctly now. If we want to provide distinct ZIM files for the different variants that should be easily doable by adding the variant parameter to the action parse api call.

**Comment by kelson42** [comment]
This issue is very important for our users. If there is a "straight" way to implement this, we should implement it. Moving to current milestone.

**Comment by Markus-Rost** [comment]
The original goal of this issue to consistently use the default mainland simplified variant for the ZIM should already be reached since we are using ActionParse. (Some ZH user please correct me if I'm wrong.)

To add ZIM files for different variants, we would just have to add the `&variant=zh-hant` parameter to the `action=parse` API request. We would only need a way to provide the desired variant as an argument. Not sure if the variant support would still be part of this issue though or if it should get it's own.

**Comment by kelson42** [comment]
@Popolechien Does the latest Wikipedia in Chinese use everywhere mainland simplified characters?

@benoit74 Do we have a norm which clearly define how to deal with multiple "alphabets" for the same content?

**Comment by Popolechien** [comment]
@kelson42 I am not sure how it works, but basically there's a toggle on-wiki that allows people to switch from one set of writing to any of 5 others. 
These being characters, there are no spelling differences as we would see in English, e.g. _Labour_/_Labor_. The closest I can think of is A->a so in a sense it is purely changing the way words are displayed, not rewriting. But I suspect a given article's default is that of the person who wrote it (e.g. Taiwanese or Mainland). Would be worth asking over at zh.wikipedia

**Comment by erickguan** [comment]
The words (comprise of some characters) are actually different in variants. This is due to regional differences.

It's useful considering simplified/traditional Chinese are a normalized form as a first pass of rendering. The source can be a mix.
The second pass is a dictionary lookup to convert words ba variants. Depending on fields, the rendering results will be very different or the same.

**Comment by kelson42** [comment]
@Popolechien  Our question is: Does https://browse.library.kiwix.org/viewer#wikipedia_zh_all_maxi_2025-05/User%3AThe_other_Kiwix_guy/Landing use simplified mainland character only?

**Comment by Popolechien** [comment]
@kelson42 this page is simplified (mainland) Chinese, yes.

**Comment by kelson42** [comment]
> @kelson42 this page is simplified (mainland) Chinese, yes.

@Popolechien To your experience do all other pages of this ZIM are using mainland simplified characters as well?

**Comment by Popolechien** [comment]
I would defer to @erickguan to spot check a few pages and confirm whether this is the case or not. But we regularly have mainland users writing to complain that a lot of text is in traditional Chinese.

**Comment by erickguan** [comment]
@Popolechien no, some traditional Chinese on the page:
- 條目
- 藝術
- 歷史
- 體育

This page has mainly simplified Chinese and some traditional Chinese https://browse.library.kiwix.org/content/wikipedia_zh_all_maxi_2025-05/%E8%AE%A1%E7%AE%97%E6%9C%BA%E7%A7%91%E5%AD%A6

This page has a mixed simplified Chinese and traditional Chinese https://browse.library.kiwix.org/viewer#wikipedia_zh_all_maxi_2025-05/%E9%90%B5%E8%B7%AF%E5%88%97%E8%BB%8A

**Comment by Markus-Rost** [comment]
Interesting, looks like just using `uselang=zh` does indeed still return a mix of both. I need to specifically set a variant. And ideally we will want to provide that variant to our loading of modules as well.

What variant would be the best one to use when the wiki language is `zh`? `zh-hans` or `zh-cn`

**Comment by benoit74** [comment]
Is there really a difference between `zh-hans` and `zh-cn` variants? I see only `zh-cn` listed on https://zh.wikipedia.org/ dropdown list / popup.

I think we should build a "demo" ZIMs of top 100 or 1k (we can use lists at https://download.openzim.org/wp1/zhwiki/tops/) before merging any PR, and proof-read it with some native / fluent Chinese reader.

**Comment by erickguan** [comment]
This depends on how you design OpenZIM.

1. zh-wikip has complex conversion tables to accommodate nuances. The relationships among these variants:

    - Simplified Chinese: zh-hans ⊃ zh-cn ⊃ zh-sg ⊃ zh-my
    - Traditional Chinese: zh-hant ⊃ zh-tw ⊃ zh-hk ⊃ zh-mo

2. If you are optimizing for languages (zh-hans/zh-hant), considering some users are not familar with character sets, this is acceptable. Then you will publish two offline sets of Chinese Wikipedia.
If you are optimizing for variants, e.g., you want users to be able to understand a movie title, a scientifical article with the words in their regions, use the detailed varitants (zh-cn, zh-sg, zh-my, zh-tw, zh-hk, zh-mo). Then you will publish 6 offline sets of Chinese Wikipedia.
   - Of course, you can consider compress the difference between variants to a generic one since they aren't too different in a diff file.
   - I only skimmed zim format so I am not entirely sure how compression can be done.

**Comment by benoit74** [comment]
Thank @erickguan for the details. 

I believe that in a first step, we should publish a single ZIM which is readable by most persons. Correct me if I'm wrong, but it seems that we must then use `zh-cn` as the **single** ZIM file we publish.

In a second step, we will have indeed have to decide if we publish 6 ZIMs, one per variant, or if we want a single ZIM with all variants inside (and whatever UI to be able to switch). This will need a more detailed analysis and discussion: impact on ZIM size if we build a single one, added value of having all variants in one single ZIM given the fact that most people have a preference, ...

**Comment by erickguan** [comment]
@benoit74 You are welcome. It's technical sensible approach to choose one. And `zh-cn` has largest potential impact. I would communicate this carefully so that people don't care about tech think this is discrimination.

And yes for the second step. It's entirely up to you to decide what to design and also the trade offs.

**Comment by benoit74** [comment]
I've moved the second step in https://github.com/openzim/mwoffliner/issues/2399 so that we do not forget about it

**Comment by Popolechien** [comment]
@benoit74 Any idea when this fix will move to PROD?

**Comment by benoit74** [comment]
It is already in PROD. Just need to wait recipe to be rescheduled within a month (or request them manually should you be in a hurry) for ZIMs to be recreated.

**Comment by Popolechien** [comment]
I have someone who wrote about this exact request and so I have requested the recipe again (whose last two iterations had failed).

**Comment by kelson42** [comment]
@Popolechien @benoit74 Anything speaking against rescheduling now all chinese based MWoffliner recipes? This is what I would do considering the bug/problem is important/impacting!

**Comment by Popolechien** [comment]
Well let's wait until someone can confirm the new WP is homogeneously simplified Chinese and then yes, let's go.

**Comment by kelson42** [comment]
@erickguan Finaly we have regenerated Wikipedia in Chinese. Could you please confirm that everything (all the content) is written with mainland simplified characters? https://browse.library.kiwix.org/viewer#wikipedia_zh_all_maxi_2025-07/User%3AThe_other_Kiwix_guy/Landing

**Comment by erickguan** [comment]
@kelson42 Confirm, splendid! Thanks for getting the variant to work.

**Comment by kelson42** [comment]
@erickguan Very very great! Thank you!

**Comment by danielzgtg** [comment]
> Is there really a difference between zh-hans and zh-cn variants?

@benoit74 Yes, there is a difference:
* https://zh.wikipedia.org/zh-hans/数量级_(时间)#奈秒（ns）：10-9秒 appears as 奈
* https://zh.wikipedia.org/zh-cn/数量级_(时间)#奈秒（ns）：10-9秒 appears as 纳

See https://zh.wikipedia.org/wiki/MediaWiki:Conversiontable/zh-cn for more details. The dropdown doesn't have zh-hans (like LANG=C in C++) because most people who would use it prefer zh-cn (like LANG=en_US.UTF-8 in C++).
